wrappers/python/openmm/app/internal/xtc_utils/src/xtc.cpp

```diff
@@ -0,0 +1,326 @@
+/*
+MIT License
+
+Copyright (c) 2023 Accellera
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+
+Contributors: Stefan Doerr, Raul P. Pelaez
+*/
+#include "xtc.h"
+#include "xdrfile.h"
+#include "xdrfile_xtc.h"
+#include <errno.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <string>
+#include <sys/stat.h>
+#include <sys/types.h>
+
+#ifndef PATH_MAX
+    #define PATH_MAX 2048
+#endif
+
+#include <sys/stat.h>
+
+static void* condfree(void* p) {
+    if (p)
+        free(p);
+    return NULL;
+}
+
+// I redefine here for frames to optimize calculations in the single frame use
+// case
+#define Xf(atom, frame, nframes) atom * 3 * nframes + frame
+#define Yf(xidx, nframes) xidx + nframes
+#define Zf(yidx, nframes) yidx + nframes
+
+int xtc_natoms(char* filename) {
+    int natoms = 0;
+    if (exdrOK != read_xtc_natoms(filename, &natoms)) {
+        fprintf(stderr, "xtc_read(): could not get natoms\n");
+        return -1;
+    }
+    return natoms;
+}
+
+int xtc_nframes(char* filename) {
+    int natoms = 0;
+    int nframes = 0;
+    rvec* p = NULL;
+    XDRFILE* xd = NULL;
+    float time;
+    int step;
+    float prec;
+    matrix box;
+    if (exdrOK != read_xtc_natoms(filename, &natoms)) {
+        fprintf(stderr, "xtc_read(): could not get natoms\n");
+        return -1;
+    }
+    if (!natoms) {
+        fprintf(stderr, "xtc_read(): natoms is 0\n");
+        return -1;
+    }
+    xd = xdrfile_open(filename, "r");
+    if (NULL == xd) {
+        fprintf(stderr, "xtc_read(): could not open file\n");
+        return -1;
+    }
+    p = (rvec*)malloc(sizeof(rvec) * natoms);
```

Comment:
Using malloc to create local workspace introduces a high risk of memory errors.  In C++, you should almost never explicitly allocate memory except in a constructor.  The correct way to do this is

```c++
vector<rvec> p(natoms);
```

You can call `p.data()` to get a pointer to the data.  The memory is guaranteed to be released correctly no matter where or how the function exits.